src/za/ac/sun/cs/green/service/simplify/ConstantPropogation.java

```diff
@@ -0,0 +1,216 @@
+package za.ac.sun.cs.green.service.simplify;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.HashMap;
+import java.util.Set;
+import java.util.Stack;
+import java.util.logging.Level;
+
+import za.ac.sun.cs.green.Green;
+import za.ac.sun.cs.green.Instance;
+import za.ac.sun.cs.green.service.BasicService;
+import za.ac.sun.cs.green.expr.Constant;
+import za.ac.sun.cs.green.expr.IntConstant;
+import za.ac.sun.cs.green.expr.Operation;
+import za.ac.sun.cs.green.expr.Variable;
+import za.ac.sun.cs.green.expr.Expression;
+import za.ac.sun.cs.green.expr.Visitor;
+import za.ac.sun.cs.green.expr.VisitorException;
+
+public class ConstantPropogation extends BasicService {
+
+	public ConstantPropogation(Green solver) {
+		super(solver);
+	}
+
+	@Override
+	public Set<Instance> processRequest(Instance instance) {
+		@SuppressWarnings("unchecked")
+		Set<Instance> result = (Set<Instance>) instance.getData(getClass());
+		if (result == null) {
+			final Expression e = simplify(instance.getFullExpression());
+			final Instance i = new Instance(getSolver(), instance.getSource(), null, e);
+			result = Collections.singleton(i);
+			instance.setData(getClass(), result);
+		}
+		return result;
+	}
+
+	public Expression simplify(Expression expression) {
+		try {
+			log.log(Level.FINEST, "Before simplification: " + expression);
+			SimplificationVisitor simplificationVisitor = new SimplificationVisitor();
+			expression.accept(simplificationVisitor);
+			Expression simplified = simplificationVisitor.getExpression();
+			log.log(Level.FINEST, "After simplification: " + simplified);
+			return simplified;
+		} catch (VisitorException x) {
+			log.log(Level.SEVERE, "Houston, we have a problem!", x);
+		}
+		return null;
+	}
+
+	private static class SimplificationVisitor extends Visitor {
+
+		private Stack<Expression> stack;
+		private Map<Expression, Constant> knownValues;
+
+		public SimplificationVisitor() {
+			stack = new Stack<Expression>();
+			knownValues = new HashMap<Expression, Constant>();
+		}
+
+		public Expression getExpression() {
+			Expression top = stack.pop();
+			top = propogateConstants(top);
+			return top;
+		}
+
+		@Override
+		public void postVisit(Variable variable) {
+			pushAndPrint(variable);
+		}
+
+		@Override
+		public void postVisit(Constant constant) {
+			pushAndPrint(constant);
+		}
+
+		@Override
+		public void postVisit(Operation operation) {
```

Comment:
Method is nice and brief, but a bit of Javadoc and maybe an inline comment or two would make it clear why this `postVisit` is so different from the other ones.